[Adreno] Add documentation for Adreno deployment (#13393)
apache/tvm: docs/how_to/deploy/adreno.rst

   support for automatic format conversions

Overall, with textures, it is possible to achieve a significant performance boost
compared to OpenCL buffer based solutions.

Building TVM for Adreno
-----------------------

This section gives instructions on how to build the Android part of TVM
with OpenCL-SDK and TVM RPC Server in order to deploy models on Adreno.

Since the process of building TVM for Adreno is exactly the same as the
process of building TVM for Android, please refer to these instructions:
`TVM RPC
Server <https://github.com/apache/tvm/tree/main/apps/cpp_rpc>`_.

Alternatively, to build a TVM via docker using OpenCL-Headers and set-up
with Android TVM RPC, refer to this guide: `Deploy the Pretrained Model on Android <https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_android.html>`_.

**Prerequisites**: Android NDK, Android Debug Bridge and OpenCL-SDK must
be installed and Android part of TVM must be built:

- Read documentation about *Android NDK installation* here: https://developer.android.com/ndk
- To get access to adb tools you can see *Android Debug Bridge installation* here: https://developer.android.com/studio/command-line/adb

> - For *OpenCL-SDK installation* please refer to official github repository: https://github.com/KhronosGroup/OpenCL-SDK.git

Reviewer: OpenCL SDK is not necessary after https://github.com/apache/tvm/pull/13362. Only one requirement: the device should have OpenCL support.

You can also build the android part of TVM locally. From the root
folder of TVM:

::

   mkdir build_android
   cd build_android

>    cmake .. -DUSE_OPENCL=path/to/OpenCL -DCMAKE_TOOLCHAIN_FILE=${ANDROID_NDK_HOME}/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_NATIVE_API_LEVEL=android-28 -DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=ON -DANDROID_STL=c++_static -DUSE_CPP_RPC=ON

Reviewer: ```suggestion
   cmake .. -DUSE_OPENCL=ON -DCMAKE_TOOLCHAIN_FILE=${ANDROID_NDK_HOME}/build/cmake/android.toolchain.cmake -DANDROID_ABI=arm64-v8a -DANDROID_NATIVE_API_LEVEL=android-28 -DCMAKE_FIND_ROOT_PATH_MODE_PACKAGE=ON -DANDROID_STL=c++_static -DUSE_CPP_RPC=ON
```

   make -jN tvm_runtime tvm_rpc

where **N** is the number of cores available on your *CPU*.

At this stage you have built TVM for Adreno.

Build and deploy model for Adreno
---------------------------------

In this section we will focus on target, needed to compile and deploy models for Adreno, demonstrate
the differences in generated kernels with and without textures and, in addition, the
possibility of choosing a different precision for model compilation will
be considered.

For the complete step-py-step process of compiling and deploying models on
Adreno, including selection of precision, running the inference of the

> model, getting the predictions, and measuring the performance please refer to this tutorial: `How To Deploy model on Adreno <https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_adreno.html>`_

Reviewer: Probably the link can be relative

      if  dtype == "float16":
         global  conv2d_acc = "float16"
         from  tvm.ir  import  IRModule
         mod = IRModule.from_expr(mod)
         seq = tvm.transform.Sequential(
            [
                  relay.transform.InferType(),
                  relay.transform.ToMixedPrecision()
            ]
         )
         with  tvm.transform.PassContext(opt_level=3):
            mod = seq(mod)
      return  mod

   dtype="float16"
   mod = convert_to_dtype(mod["main"], dtype)

We then can compile our model in any convinient way

.. code:: python

   with  tvm.transform.PassContext(opt_level=3):
       lib = relay.build(
           mod, target_host=target_host, target=target, params=params
       )

**float16_acc32 (Mixed Precision)**

ToMixedPrecision pass traverse over the network and split network to clusters of ops dealing with float or float16 data types.
The clusters are defined by three types of operations:
* Operations always be converted into float16 data type
* Operations which can be converted if they follow by converted cluster
* Operations never be converted to the float16 data type  
This list is defined in the ToMixedPrecision implementation here 

> `relay/transform/mixed_precision.py <https://github.com/apache/tvm/blob/main/python/tvm/relay/transform/mixed_precision.py#L34>`_ 

Reviewer: Probably it would be better to have relative links. Not sure if it will work with specifying line number. But anyway, line number might be not actual after adding something to this file.

Author: Couldn't make a relative link to other site.

and can be overridden by user

In some cases, we want higher precision in accumulation than the input data.
This is supported, for example, for conv2d and dense operations. To override accumulation type you need to register

> function with @register_mixed_precision_conversion decorator to modify parameters of ToMixedPrecision conversion

Reviewer: ```suggestion
function with ``@register_mixed_precision_conversion`` decorator to modify parameters of ``ToMixedPrecision`` conversion
```

Now we need to modify the conversion function by adding some logical "forks" and ToMixedPrecision() call,
then create a Relay graph from desired model in any convinient way and obtain **mod** (which is IR representation of the model),
after which we can convert it to the required **dtype** and then assemble our model sequentialy

.. code:: python

   def  convert_to_dtype(mod, dtype):
       # downcast to float16
       if  dtype == "float16"  or  dtype == "float16_acc32":
           global  conv2d_acc
           conv2d_acc = "float16"  if  dtype == "float16"  else  "float32"
           from  tvm.ir  import  IRModule
           mod = IRModule.from_expr(mod)
           seq = tvm.transform.Sequential(
               [
                   relay.transform.InferType(),
                   relay.transform.ToMixedPrecision()
               ]
           )
           with tvm.transform.PassContext(
                config={"relay.ToMixedPrecision.keep_orig_output_dtype": True},
                opt_level=3):
            mod = seq(mod)
       return  mod

   dtype="float16_acc32"
   mod = convert_to_dtype(mod["main"], dtype)
   dtype = "float32"  if  dtype == "float32"  else  "float16"

> The "ToMixedPrecision" method is a pass to convert an FP32 relay graph into an FP16 version (with

Reviewer: ```suggestion
The ``ToMixedPrecision`` method is a pass to convert an FP32 relay graph into an FP16 version (with
```

|High-level overview of the Adreno A5x architecture for OpenCL|

*Fig. 1 High-level overview of the Adreno A5x architecture for OpenCL*

*source:* `OpenCL Optimization and Best Practices for Qualcomm Adreno GPUs <https://dl.acm.org/doi/10.1145/3204919.3204935>`_

Reasons of using textures:

-  Texture processor (TP) has a dedicated L1 cache, which is read-only cache and stores data
   fetched from level-2 (L2) cache for texture operations (primary
   reason)

-  The handling of image boundaries is built-in.

-  Supports numerous image format and data type combinations with
   support for automatic format conversions

Overall, with textures, it is possible to achieve a significant performance boost
compared to OpenCL buffer based solutions.

.. _building_tvm_for_adreno:

Building TVM for Adreno
-----------------------

This section gives instructions on how to build the Android part of TVM
with OpenCL and TVM RPC Server in order to deploy models on Adreno.

Since the process of building TVM for Adreno is exactly the same as the
process of building TVM for Android, please refer to these instructions:
`TVM RPC
Server <https://github.com/apache/tvm/tree/main/apps/cpp_rpc>`_.

> Alternatively, to build a TVM via docker using OpenCL-Headers and set-up
> with Android TVM RPC, refer to this guide: `Deploy the Pretrained Model on Android <https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_android.html>`_.

Reviewer: Phrase about `OpenCL-Headers` is not actual now. OpenCL Headers are not necessary for building TVM now and this documentation also was updated in this commit: https://github.com/apache/tvm/commit/2bb3382c89419820dab51a5a0fe4f09bb7ce6ecd

```suggestion
Since there are many required packages for Android, you can use the official Docker Image to build TVM. For more information refer to this guide: `Deploy the Pretrained Model on Android <https://tvm.apache.org/docs/how_to/deploy_models/deploy_model_on_android.html>`_.
```
